resources/asciidoctor/lib/open_in_widget/extension.rb

```diff
@@ -0,0 +1,109 @@
+# frozen_string_literal: true
+
+require 'fileutils'
+
+require_relative '../scaffold.rb'
+
+##
+# Extensions for enriching certain source blocks with "OPEN IN CONSOLE",
+# "OPEN IN SENSE", "OPEN IN KIBANA", AND/OR "COPY_AS_CURL".
+#
+# Usage
+#
+#   [source,console]
+#   ---------
+#   GET /
+#   ---------
+#
+# or
+#
+#   [source,sense]
+#   ---------
+#   GET /
+#   ---------
+#
+# or
+#
+#   [source,kibana]
+#   ---------
+#   GET /
+#   ---------
+#
+# or
+#
+#   [source,sense,snippet=path/to/snippet.console]
+#   ---------
+#   GET /
+#   ---------
+#
+class OpenInWidget < TreeProcessorScaffold
+  include Asciidoctor::Logging
+
+  CALLOUT_SCAN_RX = / ?#{Asciidoctor::CalloutScanRx}/.freeze
+
+  def process_block(block)
+    return unless block.context == :listing && block.style == 'source'
+
+    lang = block.attr 'language'
+    return unless %w[console sense kibana].include? lang
+
+    snippet = block.attr 'snippet'
+    if snippet
+      logger.warn message_with_context "reading snippets from a path makes the book harder to read", :source_location => block.source_location
+      # If you specify the snippet path then we should copy it into the
+      # destination directory so it is available for Kibana.
+      snippet_path = "snippets/#{snippet}"
+      normalized = block.normalize_system_path(snippet_path, block.document.base_dir)
+      if File.readable? normalized
+        copy_snippet block, normalized, snippet_path
+      else
+        logger.warn message_with_context "can't read snippet from #{normalized}", :source_location => block.source_location
```

Comment:
Is `warn` sufficient here, or should it fail more aggressively? This looks like a pretty bad state to be in and seems like it would produce broken docs (pieces missing).

Reply:
`warn` is the level that Asciidoctor fails when it can't include a referenced file. We have to fail the asciidoctor run if it emits any warnings. Which, now that you mention them, would be another great integration test. Which I'll do in a separate PR because it'll have a bigger splash radius.

Reply:
Hmm - I've rechecked this. Asciidoctor actually spits out an `ERROR` for missing files. I can switch to that. But it still keeps going which isn't really a terrible thing because it allows you to collect all of the failures. Sometimes that results in kind of a mess, but often it is quite helpful.

Reply:
Yeah for sure that is a better behavior, we just also really want `error` when something goes as sideways as "an entire whole thing was just not there" :)

Reply:
`asciidoc` is the one that warns! asciidoctor makes an error..... I had my programs crossed.